PCR-cDNA libraries were prepared using the SQK-PCS110 PCR-cDNA-sequencing kit following the standard Oxford Nanopore Technologies (ONT) protocol24. Samples were barcoded (#1-8) and sequenced on PromethION. Two samples were loaded per flow cell (15 ng each), with paired samples run together and unpaired samples randomly split into pairs. The base calling and demultiplexing were performed live using original ONT software. Samples were processed following the previously described pipeline denoted as salmonminimap225. In brief: reads were aligned to the reference transcriptome (version GRCh38) using minimap2 (secondary score ratio -p 0.99), and the transcript abundances were estimated from the output bam files using Salmon in alignment-based mode26. Low abundant transcripts with less than 3 reads across less than 5 samples, which is 50% of the smallest study group, were removed. Transcripts corresponding to ribosomal RNA (rRNA), mitochondrial RNA and haemoglobin genes were excluded from downstream analysis. Total of 45,231 transcripts were included in the downstream analysis. Density plots of log-cpm normalized data before and after filtering are available as supplementary figure 21 (e, f).
